src/init.cpp

```diff
@@ -2058,42 +2058,67 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+        for (auto& index : node.indexes) {
+            const IndexSummary& summary = index->GetSummary();
+            if (summary.synced) continue;
+
+            // Get the last common block between the index best block and the active chain
+            const CBlockIndex* pindex = WITH_LOCK(::cs_main, return chainman.m_blockman.LookupBlockIndex(summary.best_block_hash));
+            if (!index_chain.Contains(pindex)) {
+                pindex = index_chain.FindFork(pindex);
+            }
+            if (!pindex) {
+                pindex = index_chain.Genesis();
+            }
 
-    for (auto index : node.indexes) {
-        const IndexSummary& summary = index->GetSummary();
-        if (summary.synced) continue;
+            if (index->RequiresUndoData()) {
+                if (undo_start && pindex->nHeight >= undo_start.value()->nHeight) continue;
+                undo_start = pindex;
+                if (undo_start.value()->nHeight == 0) {
+                    // We only check for undo down to the first block because Genesis can not have undo data.
+                    undo_start = index_chain[1];
+                    // Still check for block data availability of Genesis.
+                    block_start = index_chain.Genesis();
+                    block_start_name = summary.name;
+                }
+                undo_start_name = summary.name;
+            } else {
+                if (block_start && pindex->nHeight >= block_start.value()->nHeight) continue;
+                block_start = pindex;
+                block_start_name = summary.name;
+            }
+        };
 
-        // Get the last common block between the index best block and the active chain
-        LOCK(::cs_main);
-        const CBlockIndex* pindex = chainman.m_blockman.LookupBlockIndex(summary.best_block_hash);
-        if (!index_chain.Contains(pindex)) {
-            pindex = index_chain.FindFork(pindex);
+        // Verify all blocks needed to sync to current tip are present including undo data.
+        if (undo_start) {
+            LOCK(::cs_main);
+            if (!chainman.m_blockman.CheckBlockDataAvailability(*index_chain.Tip(), *Assert(undo_start.value()), BLOCK_HAVE_DATA | BLOCK_HAVE_UNDO)) {
+                return InitError(Untranslated(strprintf("%s best block of the index goes beyond pruned data (including undo data). Please disable the index or reindex (which will download the whole blockchain again)", undo_start_name)));
+            }
         }
 
-        if (!indexes_start_block || !pindex || pindex->nHeight < indexes_start_block.value()->nHeight) {
-            indexes_start_block = pindex;
-            older_index_name = summary.name;
-            if (!pindex) break; // Starting from genesis so no need to look for earlier block.
+        // Verify all blocks needed to sync to current tip are present unless we already checked all of them above.
+        if (block_start && !(undo_start && undo_start.value() <= block_start.value())) {
```

Comment:
I think you are just comparing pointers here, but actually want to compare their height. This block of code is also kind of hard to test against from a functional test. Maybe we can add a unit test for `StartIndexBackgroundSync` that manipulates some of the internal block index state?

Reply:
Right, good catch on the pointers!

I can try to add such a unit test, do you have specific scenarios in mind that you would like to see covered? If we just want to make sure this block is covered I think I could also add a functional test that basically executes the edge case that @furszy described above: https://github.com/bitcoin/bitcoin/pull/29770#discussion_r2029126932 and I think that should be fairly easy to do.